### ISSUE TITLE
Utilize the tests inside official dart repo to verify correctness?

### ISSUE THREAD

**Issue by fzyzcjy** [ethanblake4/dart_eval#60]
(Have not checked into details, just rough ideas.) Maybe we can utilize the tests inside official dart repo, so that we get a ton of test cases to verify dart-eval is correct, at least up to the dart repo :)

**Comment by ethanblake4** [comment]
Good idea. Do you know where these tests are? If not I'll poke around and see what I can find when I have time

**Comment by fzyzcjy** [comment]
I have not looked into it, but I guess there must be some

**Comment by ethanblake4** [comment]
Some good ones here: https://github.com/dart-lang/sdk/tree/main/tests/language

Would be a good project for a contributor IMO, I'll add help wanted / good first issue. 
For anyone who does want to take this on: It's important we don't end up testing the Dart language itself, just how well we match it. If Dart changes its internal implementation of the Regexp engine, that shouldn't cause us to fail as long as we're doing the same thing as Dart. Will definitely require modifying those tests to reflect that.

**Comment by marcov-dart** [comment]
It's been a while since this was posted. Did anyone take this up? If not I'll do it.

**Comment by Zverik** [comment]
That would be amazing, Marco! I looked through the tests but had absolutely no idea how to adapt them to dart eval.

**Comment by ethanblake4** [comment]
+1 - I'll assign you :)

**Comment by marcov-dart** [comment]
@Zverik 
Did some research: These tests are simply programs with some expected behaviour that they basically report using package expect like this example I found:

```
import 'package:expect/expect.dart';

main() {
  var list = [0, 1];
  Expect.throwsRangeError(() {
    list[2];
  });
}
```

There is a database with apparently 476 supported, tested configurations to compare against. Lol, what the?

So I'm thinking I am going to just try and run them through dart_eval.

**Comment by ethanblake4** [comment]
My expection (heh) is that package:expect will probably not work in dart_eval, but it's worth a shot...

**Comment by marcov-dart** [comment]
Package expect is just 1 single 130 line file. The test harnass is kept really simple on purpose. Because you don't want to have your tests depend on the thing you are trying to test too much...

**Comment by Zverik** [comment]
It's a bit [longer that 130 lines](https://github.com/dart-lang/sdk/blob/main/pkg/expect/lib/expect.dart) :)

**Comment by marcov-dart** [comment]
Yeah, I noticed. I was looking at a very old version on pub.dev. It depends on package matcher, adds two functions and exposes the whole matcher api.
I created a version of package expect and I tried to have dart_eval create a bridge. It doesn't quite work unfortunately. There are things missing in dart_eval. For instance $NoSuchMethodError and $TypeError.
In any case I think having those tests is a bit too valuable to not put the effort in. So I'll just press on and go ahead adding the missing bits.

**Comment by marcov-dart** [comment]
I have a POC here running a single set of tests in bin/dart_eval_tests.dart

https://github.com/marcov-dart/dart_eval_tests

**Comment by Zverik** [comment]
Wow. Looks good, though maybe a bit simple for a target. The next step would be cloning sources and somehow using the original files?

I also wonder if you could run all the tests instead of stopping at the first failed one, and counting the percent?

**Comment by marcov-dart** [comment]
Well, it is a work in progress and right now just a POC. Yes, I want to pull in all the (relevant) tests from the dart repo and run everything one-by-one and compare to the database with results. That is the end goal, but I suspect it is going to be a somewhat bumpy road.
Dart_eval currently is not quite able to create a working bridge to the Expect class. I have to manually edit it to make it even compile. I'm going back at some point to make that work the way it should, without manual edits. 
Next, when I run even this really, fairly simple test code, dart_eval fails to run it properly. Not in my modified version, but also not using current master. So yeah, that is an issue that needs fixing.

**Comment by marcov-dart** [comment]
So I'm running into something that has to do with the internals of the runtime. I am running the following program through dart_eval:

```dart
import "package:expect/expect.dart";

class BoolTest {
 static void testEquality() {    
    Expect.equals(true, true);
    Expect.equals(false, false);
    Expect.isTrue(identical(true, true));
    Expect.isFalse(identical(true, false));
    Expect.isTrue(identical(false, false));
    Expect.isFalse(identical(false, true));
    Expect.isFalse(!identical(true, true));
    Expect.isTrue(!identical(true, false));
    Expect.isFalse(!identical(false, false));
    Expect.isTrue(!identical(false, true));
    Expect.isTrue(true == true);
    Expect.isFalse(true == false);
    Expect.isTrue(false == false);
    Expect.isFalse(false == true);
    Expect.isFalse(true != true);
    Expect.isTrue(true != false);
    Expect.isFalse(false != false);
    Expect.isTrue(false != true);
    Expect.isTrue(identical(true, (true == true)));
    Expect.isTrue(identical(false, (true == false)));
    Expect.isTrue(identical(true, (false == false)));
    Expect.isTrue(identical(false, (false == true)));
    Expect.isFalse(!identical(true, (true == true)));
    Expect.isFalse(!identical(false, (true == false)));
    Expect.isFalse(!identical(true, (false == false)));
    Expect.isFalse(!identical(false, (false == true)));
    Expect.isFalse(identical(false, (true == true)));
    Expect.isFalse(identical(true, (true == false)));
    Expect.isFalse(identical(false, (false == false)));
    Expect.isFalse(identical(true, (false == true)));
    Expect.isTrue(!identical(false, (true == true)));
    Expect.isTrue(!identical(true, (true == false)));
    Expect.isTrue(!identical(false, (false == false)));
    Expect.isTrue(!identical(true, (false == true)));
  }

  static void testMain() {
    testEquality();
  }
}

void main() {
  BoolTest.testMain();
}
```

When I run this I get:

```

Unhandled exception:
dart_eval runtime exception: RangeError (index): Invalid value: Not in inclusive range 0..254: 255
#0      List._setIndexed (dart:core-patch/array.dart:193:48)
#1      List.[]= (dart:core-patch/array.dart:188:5)
#2      PushReturnValue.run (package:dart_eval/src/eval/runtime/ops/memory.dart:52:18)
at BoolTest.testEquality()
at BoolTest.testMain()
at main()

RUNTIME STATE
=============
Program offset: 630
Stack sample: [L249: null, L250: Instance of '$TypeImpl', L251: true, L252: ${false}, L253: ${true}, L254: ${true}]
Args sample: []
Call stack: [0, -1, 725, 719]
TRACE:
624: BoxBool (L252)
625: PushTrue ()
626: PushTrue ()
627: BoxBool (L253)
628: BoxBool (L254)
629: CheckEq (L253 == L254)
630: PushReturnValue ()  <<< EXCEPTION
631: BoxBool (L255)
632: PushArg (L252)
633: PushArg (L255)
634: InvokeExternal (Ex#1)

#0      Runtime.execute (package:dart_eval/src/eval/runtime/runtime.dart:869:7)
#1      Runtime.executeLib (package:dart_eval/src/eval/runtime/runtime.dart:842:12)
```
If I spilt the tests into two halfs then I can run both halfs. But I can't run them all at once.

**Comment by ethanblake4** [comment]
Yes, the runtime allocates stack frames of size 255. This was a temporary hack to avoid excessive allocation before building a dynamic setup and back in the day dart_eval could rarely execute functions longer than that without getting some other sort of error anyway, but seems like it'll need to be figured out pretty soon.

If you want to change it to 512 or 1024 for now go ahead (first line of run() in PushScope), shouldn't have any side effects.

**Comment by Zverik** [comment]
Yeah, I support increasing that a bit, since it was even encountered by a plugin author for my app. I guess 512 ought to be enough for anyone.

**Comment by marcov-dart** [comment]
I refactored my test code a bit and added 4000+ dart language tests and ran a first test.
287 out of 4140 are running without an exception. A great many of these tests I'm pretty sure are supposed to fail btw. And I had to exclude a whole bunch of tests because of unhandled exceptions inside dart_eval that cause execution of my program to stop despite a try/catch.

**Comment by ethanblake4** [comment]
Thanks @marcov-dart that's really cool. I would definitely accept a PR to add it once the kinks are sorted out

**Comment by marcov-dart** [comment]
Yeah, I have been thinking. My work so far shows that this can indeed work. Obviously more work is required to evaluate the output of the test programs properly, but that can be done. However, I think the easiest way to run these tests is to make use of the official dart test runner.  This test runner is designed to evaluate the dart tools. The 'old' way is to have a python script run the tests. I think it more or less is just executing "dart run" on a source file and looking at the output. I will try that first. Make an cli program with a dart_eval runtime with support for package 'expect' that mimics "dart run".

I also think that only very few of the tests will show a pass currently. I think the best aproach here is to first use the tests to identify problems that make large sections these tests fail. Then make a targeted regular tests for each of those. Fix the issues and put it up as PR.

**Comment by Zverik** [comment]
I think we would need four resulting lists: tests that fail to compile, tests that throw a runtime exception, tests that fail and tests that pass.
